### PR TITLE
feature(DecisionTree): add two-dimensional piecewise linear decision tree extending UltraRegularizedDecisionTree class

### PR DESCRIPTION
**Pull request by bugfoot**

### CODE REVIEW

File: src/finml_utils/decisiontree.py [hunk 1]
```diff
@@ -309,23 +309,254 @@ def predict(self, X: pd.DataFrame) -> pd.Series:
...
+
+
+def _calculate_bin_diff(
+    above: np.ndarray,
+    y: np.ndarray,
+    agg_method: Literal["mean", "sharpe"],
+):
+    y_below = y[~above]
+    y_above = y[above]
+    if len(y_below) != 0 and len(y_above) != 0:
```
Comment (almostintuitive) [inline src/finml_utils/decisiontree.py:534]: is it possible to have len(y_below) == 0?
as we use quantiles to define `above`

File: src/finml_utils/decisiontree.py [hunk 2]
```diff
@@ -309,23 +309,254 @@ def predict(self, X: pd.DataFrame) -> pd.Series:
...
+            X[self._endogenous_X_col], self.endogenous_thresholds_to_test, axis=0, method="closest_observation"
+        )
+        if isinstance(y, pd.Series):
+            y = y.to_numpy()
+
+        exogenous_best_split_idx = None
+        endogenous_best_split_idx = None
+        highest_abs_difference = None
+        # It could be that the best split comes from considering only the second column in X, not both.
+        if endogenous_has_variance:
```
Comment (almostintuitive) [inline src/finml_utils/decisiontree.py:395]: I think we can remove these conditionals, as none of the columns' variance can be 0

Comment (almostintuitive) [reply src/finml_utils/decisiontree.py:395]: guaranteed by the pipeline

File: src/finml_utils/decisiontree.py (hunk 1 above)
Comment (bugfoot) [reply src/finml_utils/decisiontree.py:534]: In the edge case when we have no variance in `X`. 

I will use assert in the beginning and eliminate this check.

File: src/finml_utils/decisiontree.py (hunk 2 above)
Comment (bugfoot) [reply src/finml_utils/decisiontree.py:395]: Ok, I will transform them to simple assertions.

**Pull request merged**